components/common/src/main/java/com/hotels/styx/common/FailureHandlingTaskProcessor.java

```diff
@@ -0,0 +1,144 @@
+ *
+ * It can be as strict or lenient as desired - from failing the whole run immediately as soon as a single iteration fails,
+ * to just logging failures and returning successful results.
+ *
+ * The behaviour is configured by injecting lambdas when an instance of this class is built using the {@link Builder}.
+ *
+ * @param <T> input type
+ * @param <R> output type
+ */
+public class FailureHandlingTaskProcessor<T, R> {
+    private static final Logger LOGGER = LoggerFactory.getLogger(FailureHandlingTaskProcessor.class);
+
+    private final BiConsumer<T, Exception> onEachFailure;
+    private final Consumer<Map<T, Exception>> failuresPostProcessing;
+
+    private FailureHandlingTaskProcessor(Builder<T, R> builder) {
+        this.onEachFailure = builder.onEachFailure;
+        this.failuresPostProcessing = builder.failuresPostProcessing;
+    }
+
+    /**
+     * Execute the task on a list of inputs, using the configured failure handling.
+     *
+     * @param inputs a list of inputs
+     * @param task a task to execute
+     * @return a list of outputsl
+     */
+    public List<R> process(List<T> inputs, Task<T, R> task) {
+        List<R> successes = new ArrayList<>();
+        Map<T, Exception> failures = new LinkedHashMap<>();
+
+        inputs.forEach(input -> {
+            try {
+                successes.add(task.execute(input));
+            } catch (Exception t) {
+                onEachFailure.accept(input, t);
+                failures.put(input, t);
+            }
+        });
+
+        if (!failures.isEmpty()) {
+            failuresPostProcessing.accept(failures);
+        }
+
+        return successes;
+    }
+
+    /**
+     * Builds {@link FailureHandlingTaskProcessor}.
+     *
+     * @param <T> input type
+     * @param <R> output type
+     */
+    public static final class Builder<T, R> {
+        // Logs by default so that failures do not become invisible.
+        private BiConsumer<T, Exception> onEachFailure = (input, err) -> LOGGER.error("Failed on input " + input, err);
+        private Consumer<Map<T, Exception>> failuresPostProcessing = failures -> LOGGER.error("Failures: " + failures);
+
+        /**
+         * An action to take immediately after a failure. This can be used to fail fast by rethrowing the error,
+         * or to log the failure as soon as it happens without waiting.
+         *
+         * @param onEachFailure an action to take
+         * @return this builder
+         */
+        public Builder<T, R> doImmediatelyOnEachFailure(BiConsumer<T, Exception> onEachFailure) {
+            this.onEachFailure = requireNonNull(onEachFailure);
+            return this;
+        }
+
+        /**
+         * An action to take after processing all list items, if there were failures.
+         * This will only execute if the individual failures were did not rethrow their exceptions.
+         *
+         * This action may be used to throw an exception, otherwise the {@link #process(List, Task)} method
```

Comment:
I would say that documenting what happens in 'external' code (the _otherwise_ part) should generally be avoided. You can just maybe rephrase the sentence.

Reply:
Is it really external if it is in the parent class that the builder creates?